### PR TITLE
Adding first idea for a data hole

### PR DESCRIPTION
**Pull request by hexgnu**
# Wat

This introduces a data hole for us to post to.

`POST /subscribers/:subscriber_id/data` with some JSON as a body.
# Still TODO:
- [x] I think that something like id, subscriber_id should be blacklisted from JSON but I'm curious to know what people think.
- [x] I feel like we need more testing but at the same time this doesn't really do much except create data.

### CODE REVIEW

File: app/api/made_to_measure/subscribers.rb [hunk 1]
```diff
@@ -43,6 +43,15 @@ class Subscribers < Grape::API
       subscriber
     end
 
+    desc 'Accept data to dump to a subscriber'
+    params do
+      # requires :traits, type: JSON, desc: 'JSON information'
```
Comment (hexgnu) [inline app/api/made_to_measure/subscribers.rb]: What should we validate on?

File: spec/api/made_to_measure/subscribers_spec.rb [hunk 2]
```diff
@@ -8,6 +8,7 @@
     let(:email) { 'flap@flappington.com' }
     let(:subscribers_path) { '/api/v0.1/subscribers' }
     let(:subscriber_path) { "/api/v0.1/subscribers/#{subscriber.id}" }
+    let(:subscriber_data_path) { "#{subscriber_path}/data" }
```
Comment (hexgnu) [inline spec/api/made_to_measure/subscribers_spec.rb]: Why does Grape not have path helpers? kind of annoying

File: app/api/made_to_measure/subscribers.rb (hunk 1 above)
Comment (cmhobbs) [inline app/api/made_to_measure/subscribers.rb]: Payload size?  We can't validate keys.

File: spec/api/made_to_measure/subscribers_spec.rb (hunk 2 above)
Comment (cmhobbs) [inline spec/api/made_to_measure/subscribers_spec.rb]: Agreed.  This may be something to look into when we move into MVP if we change languages.

**Comment by cmhobbs** [comment]
:cake:

**Comment by hexgnu** [comment]
reBASSed

File: Gemfile [hunk 3]
```diff
@@ -1,8 +1,10 @@
 # FIXME (cmhobbs) add hard coded versions, jerk.
 source 'https://rubygems.org'
 
+gem 'active_model_serializers'
 gem 'coffee-rails', '~> 4.1.0'
 gem 'grape'
+gem 'grape-active_model_serializers'
```
Comment (hexgnu) [inline Gemfile:7]: god made me do it

File: app/models/subscriber.rb [hunk 4]
```diff
@@ -10,4 +10,5 @@
 #
 
 class Subscriber < ActiveRecord::Base
+  has_one :subscriber_datum
```
Comment (hexgnu) [inline app/models/subscriber.rb]: what is a better name for this datum sounds too... blah to me

Comment (maxjustus) [inline app/models/subscriber.rb]: I kinda like subscriber_metadata

Comment (maxjustus) [inline app/models/subscriber.rb]: or just trash for short

Comment (maxjustus) [inline app/models/subscriber.rb]: I'm calling it metadata within the context of email templates right meow

Comment (hexgnu) [inline app/models/subscriber.rb]: `"metadata".singularize == "metadatum"` though I'll hack it in there cause it just makes more sense

**Comment by hexgnu** [comment]
K after the tests pass I think this is prolly ready to get gemerged in

File: spec/services/update_subscriber_spec.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      },
+      id: subscriber.id
+    }.as_json
+  end
+
+  it 'updates with the existing data' do
+    described_class.with_existing(params)
+
+    expect(subscriber.reload.metadata).to eq(params.fetch('metadata')
+                                            .merge(metadata))
```
Comment (cmhobbs) [inline spec/services/update_subscriber_spec.rb]: maybe store the expectation in a "result" variable?

**Comment by cmhobbs** [comment]
:cake: again

File: app/services/update_subscriber.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#
+# PATCH /subscriber/:id -> with_existing(params) means merging with existing
+# PUT /subscriber/:id -> without_existing(params) means deleting the existing
+class UpdateSubscriber
+  attr_reader :destroy_existing, :metadata, :subscriber, :subscriber_attributes
+
+  def initialize(params, destroy_existing: nil)
+    @destroy_existing = destroy_existing
+    @metadata = params.fetch('metadata')
+    @subscriber_attributes = params.reject { |k, _| k == 'metadata' }
```
Comment (maxjustus) [inline app/services/update_subscriber.rb]: just use `params.except(:metadata)`

File: app/services/update_subscriber.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    subscriber_id = params.fetch('id')
+
+    @subscriber = Subscriber.find(subscriber_id)
```
Comment (maxjustus) [inline app/services/update_subscriber.rb]: why not condense these 3 lines down to `@subscriber = Subscriber.find(params.fetch('id'))`

File: spec/services/update_subscriber_spec.rb (hunk 5 above)
Comment (maxjustus) [inline spec/services/update_subscriber_spec.rb]: totes

File: app/services/update_subscriber.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    if destroy_existing
+      subscriber.metadata = metadata
+    else
+      subscriber.metadata.merge!(metadata)
+    end
+
+    subscriber.update!(subscriber_attributes)
+  end
+
+  def self.with_existing(params)
```
Comment (maxjustus) [inline app/services/update_subscriber.rb]: I usually put `self`y methods at the top for less cronfrusion

File: app/serializers/subscriber_serializer.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# This handles basic serialization for Subscribers
+class SubscriberSerializer < ActiveModel::Serializer
+  attributes :id
```
Comment (maxjustus) [inline app/serializers/subscriber_serializer.rb]: just do `attributes :id, :name, :metadata` instead of calling `attributes` over and over

**Pull request merged**